# Exercise 4.8

## Problem 1: Imports

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Importing dataframes
path = r'C:\Users\William Inglish\Instacart Basket Analysis'

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_2.pkl'))

In [3]:
#Verifing
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id evaluation_set  order_number  order_day_of_the_week  \
0   3139998      138          prior            28                      6   
1   1977647      138          prior            30                      6   
2    389851      709          prior             2                      0   
3    652770      764          prior             1                      3   
4   1813452      764          prior             3                      4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

  _merge    price_range_loc    busiest_days busiest_period_of_day  
0   both  Mid-range product  Regularly busy           Most orders  
1   both  Mid-range product  Regularly busy        Average orders  
2   both  Mid-range product     Busiest day        Average orders  
3   both  Mid-range product      Least busy           Most orders  
4   both  Mid-range product      Least busy        Average orders

In [4]:
ords_prods_merge.shape

(32404859, 18)

## Problem 2: Aggregated Mean of Df

In [5]:
# Looking at datatypes
ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                      int64
evaluation_set              object
order_number                 int64
order_day_of_the_week        int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_days                object
busiest_period_of_day       object
dtype: object

Only aggregating the dtypes with int64 or float64.

In [6]:
#Grouping by depratment_id and looking at the mean for the order_number column using the entire dataframe

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Problem 3

In [7]:
# Creating a subset from the exercise practice.

df = ords_prods_merge[:1000000]
df.shape

(1000000, 18)

In [8]:
# Finding the aggregated mean for the order_number column in subset

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

Unsure why I get all departments, but from analyzing what  the differnces are, the subset is almost in line with the full dataframes aggregation, but there are multiple deviations either being too high or too low

## Problem 4

In [9]:
# The three step process is:
# Split the data into groups based on the “user_id” column.
# Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# Create a new column, “max_order,” into which you’ll place the results of your aggregation.

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [10]:
# Creating loyalty flags based off of max_order column

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [11]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [12]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [14]:
# Verifying the flags are correctly assigned
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number', 'max_order']].head(60)

user_id      loyalty_flag  order_number  max_order
0       138  Regular customer            28         32
1       138  Regular customer            30         32
2       709      New customer             2          5
3       764      New customer             1          3
4       764      New customer             3          3
5       777  Regular customer            16         26
6       825      New customer             3          9
7       910  Regular customer            12         12
8      1052  Regular customer            10         20
9      1052  Regular customer            15         20
10     1379      New customer             1         10
11     1480    Loyal customer            36         43
12     1480    Loyal customer            38         43
13     1480    Loyal customer            39         43
14     1494  Regular customer             6         13
15     1494  Regular customer             7         13
16     1494  Regular customer             9         13
17     1540  Regular customer             4         28
18     1540  Regular customer             5         28
19     1540  Regular customer             6         28
20     1540  Regular customer             7         28
21     1540  Regular customer             8         28
22     1540  Regular customer             9         28
23     1540  Regular customer            14         28
24     1540  Regular customer            15         28
25     1540  Regular customer            18         28
26     1540  Regular customer            19         28
27     1540  Regular customer            20         28
28     1540  Regular customer            21         28
29     1540  Regular customer            22         28
30     1540  Regular customer            24         28
31     1540  Regular customer            25         28
32     1540  Regular customer            27         28
33     1540  Regular customer            28         28
34     1598      New customer             1          4
35     1598      New customer             2          4
36     1598      New customer             3          4
37     1647      New customer             1         10
38     1647      New customer             4         10
39     1647      New customer             6         10
40     1647      New customer             8         10
41     1647      New customer             9         10
42     2833      New customer             1         10
43     2850  Regular customer             7         23
44     2857  Regular customer            24         25
45     3010  Regular customer             5         13
46     3029      New customer             1          5
47     3395      New customer             4          6
48     3904      New customer             1          3
49     4006    Loyal customer             2         42
50     4006    Loyal customer             3         42
51     4006    Loyal customer             5         42
52     4006    Loyal customer             9         42
53     4006    Loyal customer            12         42
54     4122  Regular customer            11         40
55     4141      New customer             2          9
56     4343      New customer             1          4
57     4553      New customer             3          4
58     4845  Regular customer             6         14
59     6866      New customer             1          6

## Problem 5

In [15]:
# We want to view all statistics of the loyalty category flags. This means we will group by loyalty_flag column and aggregate the mean,
# the min, and the max.

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Looking at these statistics, it is intereasting to see the prices for the longer customers decrease while the newer customers had a higher average of pricing. However, the maximum prices indicated that there might be deviation in the dataset and cleaning is needed for further analysis.

In [16]:
ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

The Max is a deviation indicated by the standard deviation being 495.655 which means we need to check the data more clearly.

In [17]:
# checkng for max values
ords_prods_merge['prices'].sort_values(ascending = False)

21786769    99999.0
21786633    99999.0
21786624    99999.0
21786625    99999.0
21786626    99999.0
             ...   
15016689        1.0
15016688        1.0
15016687        1.0
15016686        1.0
3961580         1.0
Name: prices, Length: 32404859, dtype: float64

In [18]:
# assigning temp df having all rows with price = 99999
price_dev = ords_prods_merge.loc[ords_prods_merge['prices'].isin([99999])]

price_dev

product_id           product_name  aisle_id  department_id   prices  \
21786183       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786184       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786185       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786186       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786187       33664  2 % Reduced Fat  Milk        84             16  99999.0   
...              ...                    ...       ...            ...      ...   
21786876       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786877       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786878       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786879       33664  2 % Reduced Fat  Milk        84             16  99999.0   
21786880       33664  2 % Reduced Fat  Milk        84             16  99999.0   

          order_id  user_id evaluation_set  order_number  \
21786183    183964      873          prior             3   
21786184   1851256      873          prior             4   
21786185   1915696     1893          prior             1   
21786186   2763293     1893          prior             2   
21786187   2564805     1893          prior             4   
...            ...      ...            ...           ...   
21786876   2249946   204099          prior            29   
21786877   2363282   204099          prior            31   
21786878   3181945   204395          prior            13   
21786879   2486215   205227          prior             7   
21786880   2232716   205227          prior            10   

          order_day_of_the_week  order_hour_of_day  days_since_prior_order  \
21786183                      0                 10                     7.0   
21786184                      6                 12                    13.0   
21786185                      5                 17                     NaN   
21786186                      4                 16                    13.0   
21786187                      1                 17                    30.0   
...                         ...                ...                     ...   
21786876                      0                  8                     4.0   
21786877                      0                  9                     2.0   
21786878                      3                 15                     8.0   
21786879                      3                 20                     4.0   
21786880                      0                 11                     4.0   

          add_to_cart_order  reordered _merge     price_range_loc  \
21786183                 11          0   both  High-range product   
21786184                  8          1   both  High-range product   
21786185                 10          0   both  High-range product   
21786186                  6          1   both  High-range product   
21786187                  3          1   both  High-range product   
...                     ...        ...    ...                 ...   
21786876                  1          0   both  High-range product   
21786877                  1          1   both  High-range product   
21786878                 25          0   both  High-range product   
21786879                  8          0   both  High-range product   
21786880                 23          1   both  High-range product   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
21786183     Busiest day           Most orders          8      New customer  
21786184  Regularly busy           Most orders          8      New customer  
21786185  Regularly busy        Average orders          6      New customer  
21786186      Least busy           Most orders          6      New customer  
21786187     Busiest day        Average orders          6      New customer  
...                  ...                   ...        ...               ...  
2178

It looks like product ID 33664 is the only value with deviation. I am going to replace the rows with this price to 9.99 assuming that it was human error.

In [19]:
# Replacing the rows associated with the price deviation
ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({99999: 9.99})

ords_prods_merge.shape

(32404859, 20)

In [20]:
# Rechecking the stats after cleaning
ords_prods_merge.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  9.826464e+00  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  1.741430e+02  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  1.490000e+04  3.421083e+06   

            user_id  order_number  order_day_of_the_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07           3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01           2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01           2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00           0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00           1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01           3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01           5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01           6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  
count            3.032876e+07       3.240486e+07  3.240486e+07  3.240486e+07  
mean             1.110408e+01       8.352547e+00  5.895873e-01  3.305217e+01  
std              8.779064e+00       7.127071e+00  4.919087e-01  2.515525e+01  
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00  
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01  
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01  
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.700000e+01  
max              3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01

It looks like it is better, but the price is too high still for it being 14900.0 I will check the deviation and modify if it is necessary

In [21]:
# assigning temp df to the rows with the deviation
price_dev = ords_prods_merge.loc[ords_prods_merge['prices'].isin([14900])]

price_dev

product_id                      product_name  aisle_id  \
13100147       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13100148       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13100149       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13100150       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13100151       21553  Lowfat 2% Milkfat Cottage Cheese       108   
...              ...                               ...       ...   
13104571       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13104572       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13104573       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13104574       21553  Lowfat 2% Milkfat Cottage Cheese       108   
13104575       21553  Lowfat 2% Milkfat Cottage Cheese       108   

          department_id   prices  order_id  user_id evaluation_set  \
13100147             16  14900.0    912404       17          prior   
13100148             16  14900.0    603376       17          prior   
13100149             16  14900.0   3264360      135          prior   
13100150             16  14900.0    892534      135          prior   
13100151             16  14900.0    229704      342          prior   
...                 ...      ...       ...      ...            ...   
13104571             16  14900.0   3172853   205650          prior   
13104572             16  14900.0   2504315   205818          prior   
13104573             16  14900.0   1108388   205818          prior   
13104574             16  14900.0   1916142   206049          prior   
13104575             16  14900.0    379732   206049          prior   

          order_number  order_day_of_the_week  order_hour_of_day  \
13100147            12                      2                 14   
13100148            22                      6                 16   
13100149             2                      2                 21   
13100150             3                      0                  8   
13100151             8                      1                 19   
...                ...                    ...                ...   
13104571            18                      1                  9   
13104572             3                      5                 15   
13104573             5                      4                  5   
13104574             1                      2                 17   
13104575             4                      1                 14   

          days_since_prior_order  add_to_cart_order  reordered _merge  \
13100147                     5.0                  5          0   both   
13100148                     4.0                  3          1   both   
13100149                    13.0                  6          0   both   
13100150                    12.0                  3          1   both   
13100151                    30.0                  9          0   both   
...                          ...                ...        ...    ...   
13104571                     7.0                 17          1   both   
13104572                     3.0                 13          0   both   
13104573                     1.0                  5          1   both   
13104574                     NaN                  2          0   both   
13104575                     5.0                  4          1   both   

             price_range_loc    busiest_days busiest_period_of_day  max_order  \
13100147  High-range product  Regularly busy           Most orders         40   
13100148  High-range product  Regularly busy           Most orders         40   
13100149  High-range product  Regularly busy        Average orders          4   
13100150  High-range product     Busiest day        Average orders          4   
13100151  High-range product     Busiest day        Average orders         16   
...                      ...             ...                   ...        ...   
13104571  High-range product     Busiest day           Most orders         25   
13104572  

In [22]:
#Checking stats of this product
price_dev.describe()

product_id  aisle_id  department_id   prices      order_id  \
count      4429.0    4429.0         4429.0   4429.0  4.429000e+03   
mean      21553.0     108.0           16.0  14900.0  1.716950e+06   
std           0.0       0.0            0.0      0.0  9.830360e+05   
min       21553.0     108.0           16.0  14900.0  4.300000e+01   
25%       21553.0     108.0           16.0  14900.0  8.678360e+05   
50%       21553.0     108.0           16.0  14900.0  1.714277e+06   
75%       21553.0     108.0           16.0  14900.0  2.572842e+06   
max       21553.0     108.0           16.0  14900.0  3.420827e+06   

             user_id  order_number  order_day_of_the_week  order_hour_of_day  \
count    4429.000000   4429.000000            4429.000000        4429.000000   
mean   105002.918492     15.568526               2.727478          13.514789   
std     59094.071587     17.160597               2.015750           4.152531   
min        17.000000      1.000000               0.000000           0.000000   
25%     56669.000000      4.000000               1.000000          10.000000   
50%    103578.000000      9.000000               3.000000          14.000000   
75%    156411.000000     20.000000               5.000000          16.000000   
max    206049.000000     99.000000               6.000000          23.000000   

       days_since_prior_order  add_to_cart_order    reordered    max_order  
count             4120.000000        4429.000000  4429.000000  4429.000000  
mean                11.963835           8.201174     0.614134    30.916460  
std                  9.083407           6.922199     0.486854    25.632774  
min                  0.000000           1.000000     0.000000     3.000000  
25%                  5.000000           3.000000     0.000000    12.000000  
50%                  8.500000           6.000000     1.000000    22.000000  
75%                 17.000000          11.000000     1.000000    43.000000  
max                 30.000000          72.000000     1.000000    99.000000

In [23]:
# I will change the price from 14900 to 14.9 assuming it was human error

ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({14900: 14.90})
ords_prods_merge.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  7.792013e+00  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.242300e+00  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01  3.421083e+06   

            user_id  order_number  order_day_of_the_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07           3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01           2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01           2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00           0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00           1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01           3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01           5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01           6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  
count            3.032876e+07       3.240486e+07  3.240486e+07  3.240486e+07  
mean             1.110408e+01       8.352547e+00  5.895873e-01  3.305217e+01  
std              8.779064e+00       7.127071e+00  4.919087e-01  2.515525e+01  
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00  
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01  
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01  
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.700000e+01  
max              3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01

These results look better, now we will check on the stats of the loyalty flags again to confirm.

In [24]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.774457  1.0  25.0
New customer      7.802354  1.0  25.0
Regular customer  7.799319  1.0  25.0

Looking at these observations, the average prices of each loyalty flag have decreased to comparable measures against the full
dataframe and the max prices are aligned with the full data frame with less standard deviation.

## Problem 6

The criteria is:

    [1. If mean(prices) < 10, then flag = "Low spender"], 
    [2. If mean(prices) => 10, then flag = " High spender"]

In [25]:
# Getting a column to find the averages of each user
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id evaluation_set  order_number  order_day_of_the_week  ...  \
0   3139998      138          prior            28                      6  ...   
1   1977647      138          prior            30                      6  ...   
2    389851      709          prior             2                      0  ...   
3    652770      764          prior             1                      3  ...   
4   1813452      764          prior             3                      4  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                     3.0                  5          0    both   
1                    20.0                  1          1    both   
2                     6.0                 20          0    both   
3                     NaN                 10          0    both   
4                     9.0                 11          1    both   

     price_range_loc    busiest_days busiest_period_of_day max_order  \
0  Mid-range product  Regularly busy           Most orders        32   
1  Mid-range product  Regularly busy        Average orders        32   
2  Mid-range product     Busiest day        Average orders         5   
3  Mid-range product      Least busy           Most orders         3   
4  Mid-range product      Least busy        Average orders         3   

       loyalty_flag average_price  
0  Regular customer      6.935811  
1  Regular customer      6.935811  
2      New customer      7.930208  
3      New customer      4.972414  
4      New customer      4.972414  

[5 rows x 21 columns]

In [26]:
#Creating spending flags on criteria given using the average_price column
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [28]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >=10, 'spending_flag'] = 'High spender'

In [32]:
# Verifying by looking at the frequency of each flag.
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32284898
High spender      119961
Name: spending_flag, dtype: int64

## Problem 7

The criteria is:

    [1. If median(days since prior order) > 20, then flag = "Non-frequent customer"], 
    [2. If median(days since prior order) >10 AND <= 20, then flag = "Regular customer"],
    [3. If median(days since prior order) < 10, then flag = "Frequent customer"],

In [33]:
# Creating a column of median days since prior order grouping by user_id and transforming using median

ords_prods_merge['median_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id evaluation_set  order_number  order_day_of_the_week  ...  \
0   3139998      138          prior            28                      6  ...   
1   1977647      138          prior            30                      6  ...   
2    389851      709          prior             2                      0  ...   
3    652770      764          prior             1                      3  ...   
4   1813452      764          prior             3                      4  ...   

   reordered  _merge    price_range_loc    busiest_days busiest_period_of_day  \
0          0    both  Mid-range product  Regularly busy           Most orders   
1          1    both  Mid-range product  Regularly busy        Average orders   
2          0    both  Mid-range product     Busiest day        Average orders   
3          0    both  Mid-range product      Least busy           Most orders   
4          1    both  Mid-range product      Least busy        Average orders   

  max_order      loyalty_flag average_price  spending_flag median_prior_order  
0        32  Regular customer      6.935811    Low spender                8.0  
1        32  Regular customer      6.935811    Low spender                8.0  
2         5      New customer      7.930208    Low spender                8.0  
3         3      New customer      4.972414    Low spender                9.0  
4         3      New customer      4.972414    Low spender                9.0  

[5 rows x 23 columns]

In [34]:
# Creating order_frequency_flag based on criteria

ords_prods_merge.loc[ords_prods_merge['median_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [36]:
ords_prods_merge.loc[(ords_prods_merge['median_prior_order'] > 10) & (ords_prods_merge['median_prior_order']<= 20), 'order_frequency_flag'] = 'Regular customer'

In [37]:
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [38]:
# Checking the frequency of the new flags

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [39]:
# Checking the null values

null_order = ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()]

null_order

product_id                                       product_name  \
6234909        10749                            Organic Red Bell Pepper   
12947653       21334                              Organic Peeled Garlic   
13839012       22198  4X Ultra Concentrated Natural Laundry Detergen...   
14758536       23695                           California Veggie Burger   
21673807       33401                               Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id evaluation_set  \
6234909         83              4     5.7    895835   159838          prior   
12947653       123              4    10.2    895835   159838          prior   
13839012        75             17     1.7    895835   159838          prior   
14758536        42              1     4.7    895835   159838          prior   
21673807        21             16    14.8    895835   159838          prior   

          order_number  order_day_of_the_week  ...  _merge    price_range_loc  \
6234909              1                      0  ...    both  Mid-range product   
12947653             1                      0  ...    both  Mid-range product   
13839012             1                      0  ...    both  Low-range product   
14758536             1                      0  ...    both  Low-range product   
21673807             1                      0  ...    both  Mid-range product   

          busiest_days  busiest_period_of_day max_order  loyalty_flag  \
6234909    Busiest day         Average orders         1  New customer   
12947653   Busiest day         Average orders         1  New customer   
13839012   Busiest day         Average orders         1  New customer   
14758536   Busiest day         Average orders         1  New customer   
21673807   Busiest day         Average orders         1  New customer   

         average_price spending_flag  median_prior_order order_frequency_flag  
6234909           7.42   Low spender                 NaN                  NaN  
12947653          7.42   Low spender                 NaN                  NaN  
13839012          7.42   Low spender                 NaN                  NaN  
14758536          7.42   Low spender                 NaN                  NaN  
21673807          7.42   Low spender                 NaN                  NaN  

[5 rows x 24 columns]

Looks like user 159838 is a new user, so I will keep them in the dataframe.

## Problem 8 and 9

In [ ]:
# Exporting dataframe. My notebook is clean and structured.

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_3.pkl'))